From 8a9b4afc9bde9eb69536cddd0cd06de3be965127 Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:39:19 -0400
Subject: [ethernet] i40e: warn on newer/older firmware API rev

Message-id: <1409658042-22689-199-git-send-email-sassmann@redhat.com>
Patchwork-id: 90393
O-Subject: [RHEL7.1 PATCH 198/281] i40e: warn on newer/older firmware API rev
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Shannon Nelson <shannon.nelson@intel.com>

If the firmware's API minor number is larger than the one expected, log
a warning and recommend driver SW update.
If the firmware's API major or minor number is smaller then the one expected
(n for major, n or n-1 for minor), log a warning and recommend NVM update.

Change-ID: If0b887e055478f8e435ba7fa28113b63a6f1bb35
Signed-off-by: Shannon Nelson <shannon.nelson@intel.com>
Signed-off-by: Catherine Sullivan <catherine.sullivan@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit 278b6f629c8221b06850f43c748d0888d5d15a24)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_main.c b/drivers/net/ethernet/intel/i40e/i40e_main.c
index a148917..2a91e67 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_main.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_main.c
@@ -8633,6 +8633,20 @@ static int i40e_probe(struct pci_dev *pdev, const struct pci_device_id *ent)
   goto err_pf_reset;
  }
 
+ if (hw->aq.api_min_ver > I40E_FW_API_VERSION_MINOR)
+  dev_info(&pdev->dev,
+    "Note: FW API version %02x.%02x newer than expected %02x.%02x, recommend driver update.\n",
+    hw->aq.api_maj_ver, hw->aq.api_min_ver,
+    I40E_FW_API_VERSION_MAJOR, I40E_FW_API_VERSION_MINOR);
+
+ if (hw->aq.api_maj_ver < I40E_FW_API_VERSION_MAJOR ||
+     hw->aq.api_min_ver < (I40E_FW_API_VERSION_MINOR-1))
+  dev_info(&pdev->dev,
+    "Note: FW API version %02x.%02x older than expected %02x.%02x, recommend nvm update.\n",
+    hw->aq.api_maj_ver, hw->aq.api_min_ver,
+    I40E_FW_API_VERSION_MAJOR, I40E_FW_API_VERSION_MINOR);
+
+
  i40e_verify_eeprom(pf);
 
  /* Rev 0 hardware was never productized */
-- 
1.7.1